### PR TITLE
Feature/android

### PR DESCRIPTION
**Pull request by snodnipper**
draft android code

### CODE REVIEW

File: search-android/build.gradle [hunk 1]
```diff
@@ -0,0 +1,57 @@
+buildscript {
+    repositories {
+        jcenter()
+    }
+    dependencies {
+        classpath 'com.android.tools.build:gradle:2.0.0-beta6'
+        //classpath 'com.jakewharton.sdkmanager:gradle-plugin:0.12.0'
```
Comment (drhaynes) [inline search-android/build.gradle]: Do we need this in here?

File: search-android/src/main/java/uk/os/elements/search/android/app/SearchActivity.java [hunk 2]
```diff
@@ -0,0 +1,639 @@
...
+import java.net.UnknownHostException;
+import java.util.ArrayList;
+import java.util.List;
+
+import retrofit2.adapter.rxjava.HttpException;
+import uk.os.elements.search.android.R;
+import uk.os.elements.search.android.providers.bng.GridReferenceProvider;
+import uk.os.elements.search.android.providers.latlon.LatLonProvider;
+import uk.os.elements.search.android.providers.localdatabase.LocalDatabaseProvider;
+//import retrofit2.RetrofitError;
```
Comment (drhaynes) [inline search-android/src/main/java/uk/os/elements/search/android/app/SearchActivity.java]: Commented import, presumably can go.

File: search-android/src/main/java/uk/os/elements/search/android/app/SearchActivity.java [hunk 3]
```diff
@@ -0,0 +1,639 @@
...
+            @Override
+            public void onTextChanged(CharSequence text, int start, int before, int count) {
+                mQueryConcern.setSearchText(text.toString());
+                updateSearchViewState();
+            }
+
+            @Override
+            public void afterTextChanged(Editable s) {
+
+            }
```
Comment (drhaynes) [inline search-android/src/main/java/uk/os/elements/search/android/app/SearchActivity.java]: Can before/after overrides be removed as they aren't doing anything?

Comment (snodnipper) [inline search-android/src/main/java/uk/os/elements/search/android/app/SearchActivity.java]: sadly not as those methods are abstract http://developer.android.com/reference/android/text/TextWatcher.html

File: search-android/src/main/java/uk/os/elements/search/android/app/SearchAdapter.java [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+            }
+        });
+
+        String text = searchResult.getName();
+        String highlight = mQuery;
+
+        if (searchResult instanceof OsGridReference) {
+            highlight = searchResult.getName();
+        } else if (searchResult instanceof LatLonResult) {
+            highlight = searchResult.getName();
```
Comment (drhaynes) [inline search-android/src/main/java/uk/os/elements/search/android/app/SearchAdapter.java]: Looks like these are doing the same thing...

File: search-android/src/main/java/uk/os/elements/search/android/recentmanager/impl/provider/RecentsManagerImpl.java [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+                Double minY = cursor.getDouble(cursor.getColumnIndex(SearchResultColumns.MINY));
+                Double maxX = cursor.getDouble(cursor.getColumnIndex(SearchResultColumns.MAXX));
+                Double maxY = cursor.getDouble(cursor.getColumnIndex(SearchResultColumns.MAXY));
+                return new SearchResult(featureId, name, context, new Point(x, y),
+                        new Envelope(minX, minY, maxX, maxY), SpatialReference.create(srid));
+            }
+        }).filter(new Func1<SearchResult, Boolean>() {
+            @Override
+            public Boolean call(SearchResult result) {
+                // TODO: improve
```
Comment (drhaynes) [inline search-android/src/main/java/uk/os/elements/search/android/recentmanager/impl/provider/RecentsManagerImpl.java]: Bit more detail re: improvement would be good here. Or maybe document elsewhere...

File: search-android/src/main/java/uk/os/elements/search/android/recentmanager/impl/provider/RecentsManagerImpl.java [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+            @Override
+            public void call(Subscriber<? super Void> subscriber) {
+                SearchResultContentValues values = getValues(searchResult);
+                SearchResultSelection selection = new SearchResultSelection();
+                selection.featureid(searchResult.getId());
+                values.update(mContext.getContentResolver(), selection);
+            }
+        })
+                .subscribeOn(Schedulers.newThread())
+                .subscribe(subject);
```
Comment (drhaynes) [inline search-android/src/main/java/uk/os/elements/search/android/recentmanager/impl/provider/RecentsManagerImpl.java]: Indentation

File: search-android/src/main/java/uk/os/elements/search/android/recentmanager/impl/provider/content/RecentsProvider.java [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+import java.util.Arrays;
+
+import android.content.ContentValues;
+import android.content.UriMatcher;
+import android.database.Cursor;
+import android.database.sqlite.SQLiteOpenHelper;
+import android.net.Uri;
+
+//TODO import uk.os.elements.search.android.BuildConfig;
```
Comment (drhaynes) [inline search-android/src/main/java/uk/os/elements/search/android/recentmanager/impl/provider/content/RecentsProvider.java]: You have done, two lines below :-)

File: search-android/src/main/java/uk/os/elements/search/android/recentmanager/impl/provider/content/SearchRecentsSQLiteOpenHelper.java [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+    private SearchRecentsSQLiteOpenHelper(Context context, DatabaseErrorHandler errorHandler) {
+        super(context, DATABASE_FILE_NAME, null, DATABASE_VERSION, errorHandler);
+        mContext = context;
+        mOpenHelperCallbacks = new SearchRecentsSQLiteOpenHelperCallbacks();
+    }
+
+
+    @Override
+    public void onCreate(SQLiteDatabase db) {
+        //if (BuildConfig.DEBUG) Timber.d("onCreate");
```
Comment (drhaynes) [inline search-android/src/main/java/uk/os/elements/search/android/recentmanager/impl/provider/content/SearchRecentsSQLiteOpenHelper.java]: Commented

**Comment by drhaynes** [comment]
Looking good. All my comments are minor stuff.

**Comment by snodnipper** [comment]
Excellent, thanks Dave for your feedback.  I have rebased to remove CI config issues (known issue in android tools).

From here it would be good to look at publishing to maven central and creating some screenshots similar to iOS.

**Comment by snodnipper** [comment]
All renamed and building\* - I'll check out the aar deployment process first thing tomorrow.
- yes android studio did trash my project - console ftw

**Comment by drhaynes** [comment]
Merging for release

**Pull request merged**